### DISCUSSION TITLE
Integrate with fidget.nvim for better notification?

### DISCUSSION THREAD

**Discussion by spywhere** [wojciech-kulik/xcodebuild.nvim#18] [Ideas]
Hi, thank you for this amazing plugin. Though with my setup, the current notification is a bit intrusive as I have `cmdheight` set to `0`. It would be better if this plugin can integrated with [fidget.nvim](https://github.com/j-hui/fidget.nvim) to give a better notification experience.

fidget.nvim support both notification (a one-off message) and progress reporting, in which both can be incorporate into xcodebuild.nvim

I did have my own setup that already integrate xcodebuild.nvim with fidget.nvim but I think the progress reporting can be better if this was setup natively from xcodebuild.nvim side.

https://github.com/spywhere/dotfiles/blob/9035112dde8841f58a14176559ce938ddd6aad53/configs/nvim/lua/plugin/language/ios.lua#L35-L64

https://github.com/wojciech-kulik/xcodebuild.nvim/assets/1087399/19bc671f-f51e-4e05-b2ab-2e3c64ecc732

**Comment by wojciech-kulik** [comment]
Thank you for the idea! 

In general, I want to keep this plugin independent from other dependencies,  as much as possible. That's why I provide those functions to allow you to adjust notifications to your needs.

However, probably I could add somewhere information how to configure xcodebuild.nvim to use fidget, will think about it. Thank you 🍻

**Reply by spywhere** [reply]
I'm all for it. And it would be great if xcodebuild.nvim can expose its own progress so we can pass along to fidget to display properly. Right now it used internally inside `notifications.lua`.

**Reply by wojciech-kulik** [reply]
@spywhere thank you for suggesting this cool plugin. I will think about how I could take advantage of its features and either integrate them into the plugin or expose functions that will allow nice integration.

Also, I started thinking about using the spinner provided by Fidget and try building a window with a tests tree (like in Xcode)

**Reply by wojciech-kulik** [reply]
@spywhere I added custom auto commands. Now it should let prepare a better integration with fidget :)
https://github.com/wojciech-kulik/xcodebuild.nvim?tab=readme-ov-file#-customize-behaviors

**Reply by spywhere** [reply]
Really nice! Thank you for the update, I'll try integrate those into my setup soon 😄

**Comment by wojciech-kulik** [comment]
@spywhere I provided auto commands to notify about important plugin events (https://github.com/wojciech-kulik/xcodebuild.nvim?tab=readme-ov-file#-customize-behaviors), so now the integration should be easier and more customizable :)

I also implemented integration with this plugin in Fidget (https://github.com/j-hui/fidget.nvim/pull/212) to avoid collisions with Test Explorer 🔥 

I leave this discussion open to people who are looking for some nice integrations. Maybe in the future, I will create a wiki and then this discussion will be closed :).

If you or someone else create a nice configuration to integrate messages from xcodebuild.nvim, feel free to share it here! 🚀

**Comment by wojciech-kulik** [comment]
Here you can find my config, it works quite well:

Fidget plugin itself:
```lua
return {
  "j-hui/fidget.nvim",
  event = "VeryLazy",
  config = function()
    local fidget = require("fidget")
    fidget.setup({
      notification = {
        window = {
          normal_hl = "String", -- Base highlight group in the notification window
          winblend = 0, -- Background color opacity in the notification window
          border = "rounded", -- Border around the notification window
          zindex = 45, -- Stacking priority of the notification window
          max_width = 0, -- Maximum width of the notification window
          max_height = 0, -- Maximum height of the notification window
          x_padding = 1, -- Padding from right edge of window boundary
          y_padding = 1, -- Padding from bottom edge of window boundary
          align = "bottom", -- How to align the notification window
          relative = "editor", -- What the notification window position is relative to
        },
      },
    })
  end,
}
```

`xcodebuild.nvim` setup options:

```lua
show_build_progress_bar = false,
logs = {
  notify = function(message, severity)
    local fidget = require("fidget")
    if progress_handle then
      progress_handle.message = message
      if not message:find("Loading") then
        progress_handle:finish()
        progress_handle = nil
        if vim.trim(message) ~= "" then
          fidget.notify(message, severity)
        end
      end
    else
      fidget.notify(message, severity)
    end
  end,
  notify_progress = function(message)
    local progress = require("fidget.progress")

    if progress_handle then
      progress_handle.title = ""
      progress_handle.message = message
    else
      progress_handle = progress.handle.create({
        message = message,
        lsp_client = { name = "xcodebuild.nvim" },
      })
    end
  end,
},
```

top of the file with xcodebuild.nvim plugin:
```lua
local progress_handle
```

Demo:

https://github.com/wojciech-kulik/xcodebuild.nvim/assets/3128467/ab085ce9-ad9c-4b54-87d1-32fed4d1096f

> [!TIP]
> Key messages like build/test finished are published as notification, so if you miss one, you can check out `:Fidget history`.

**Comment by wojciech-kulik** [comment]
I'm creating a collection of useful things, so I moved this config to another discussion: #23.
Hopefully, this one resolves your request :)
